### PR TITLE
refactor: Upgrade to clap v4

### PR DESCRIPTION
**Pull request by epage**
This changes the format of help output
- Default features were enabled as we ` bat` doesn't seem to disable any and it is easier to keep up with new default features like `help`, `usage`, and `error-context`
- Help styling was removed because styling can't be applied to `help_template` atm which is required for showing `cache` command last
  - See https://github.com/clap-rs/clap/issues/1433
- Title Case is used instead of ALL CAPS
- "Command" is used instead of "Subcommand"
- Name/version isn't in `--help` anymore

### CODE REVIEW

**Comment by sharkdp** [comment]
Thank you very much!

It's probably good to mention that this changes the layout/look of the help text quite dramatically. Also, something seems wrong with the "Usage" line?

image

**Comment by epage** [comment]
> The "Use '--help' instead of '-h' to see a more detailed version of the help text." and "Note: bat -h prints a short and concise overview while bat --help gives all details." parts are missing. I would like to have them back.

Before doing so, I wanted to double check that this request is with the understanding that `--help`'s description includes this kind of context as well and that you still want the other explanations on top of that.

**Comment by sharkdp** [comment]
> > The "Use '--help' instead of '-h' to see a more detailed version of the help text." and "Note: bat -h prints a short and concise overview while bat --help gives all details." parts are missing. I would like to have them back.
> 
> Before doing so, I wanted to double check that this request is with the understanding that `--help`'s description includes this kind of context as well and that you still want the other explanations on top of that.

Not sure I understand correctly. I think that *"Use '--help' instead of '-h' to see a more detailed version of the help text."* line at the very end of the `-h` and `--help` text was very helpful, because we hide a lot of options in the `-h` text and use much shorter descriptions. I'm okay with removing the other part (which was positioned in the top description of the `-h` text).

**Comment by epage** [comment]
Clap's auto-generated `--help` flag now includes a hint about this as well, inspired by `bat`:
image
(`use --help for more details` which showed up because you used `-h`)

**Comment by sharkdp** [comment]
> Clap's auto-generated `--help` flag now includes a hint about this as well, inspired by `bat`:

Ohhh, I completely missed that. In that case.. let's hope that others don't miss it as well. And skip those additional explanations. 

Thank you for the updates! I would like to hear the opinions of the other maintainers before merging this.

**Review by sharkdp — APPROVED** [review]

**Review by keith-hall — APPROVED** [review]

File: Cargo.toml
Comment (Enselic) [inline Cargo.toml:80]: I think it would be nice to bump to latest already now (4.0.10 as of now)

**Review by Enselic — COMMENTED** [review]
First of all, thanks a lot for working on this.

I do however wonder if want clap v4 enough to sacrifice readability of help output (see separate comment).

File: src/bin/bat/clap_app.rs
Comment (Enselic) [reply src/bin/bat/clap_app.rs:55]: My concern is basically that headers stopped being orange and args stopped being yellow. Not sure what it takes to get those colors back.

Comment (epage) [reply src/bin/bat/clap_app.rs:55]: Clap v4 moved away from yellow section headers and green args.  See [the announcement for details](https://epage.github.io/blog/2022/09/clap4/#polishing-help-output).  They were replaced by bold and underline.

However, that is not why `disable_colored_help` is set.  clap v4 also moved the subcommands to the top which is great for most applications where the focus is on the subcommands but that doesn't work well for bat where the subcommand isn't a focus.  Bat now has to provide its own `help_template` to override the help ordering but this means there are "Arguments", "Options", and "Commands" headers that cannot be styled at this time but other headers are styled.  To make things consistent, I've turned off all styling.

So if you want theming from clap (https://github.com/clap-rs/clap/issues/3234), you first need `help_template` styling (https://github.com/clap-rs/clap/issues/1433).  My initial focus is on `help_template` styling.  After that I'll focus on theming.  The main issue around theming is having a stable API for specifying ansi codes.  I've created [anstyle](https://docs.rs/anstyle/latest/anstyle/) for this but need more experience reports / feedback on it before I feel its ready to go 1.0.  My hope is the `help_tempalate` work will give me experience with it internal to clap to help.

If people want to help towards check, I'd recommend checking out [anstyle](https://docs.rs/anstyle/latest/anstyle/) and providing any feedback on it as a general crate for other crates to have theming support in their API without being tied to termcolor or one of the other crates.

**Comment by ofek** [comment]
This looks great!

File: Cargo.toml
Comment (sharkdp) [reply Cargo.toml:80]: Done. (Note that the precise version is specified in the lockfile, not here)

File: src/bin/bat/clap_app.rs
Comment (sharkdp) [reply src/bin/bat/clap_app.rs]: To resolve this, I made the following changes:
- I removed the custom template, which re-enables bold/underline styling. Which takes some getting used-to (if you are used to the colorized output), but is IMO also a very reasonable styling alternative that is easy to read.
- I set the `bat cache` subcommand to hidden. This removes the whole "command" section from both `-h` and `--help` text.
- I added a new `.after_help(…)` text that let's users know that there is a `bat cache` subcommand.

Comment (Enselic) [reply src/bin/bat/clap_app.rs]: Sounds reasonable to me, nice job 👍
(Long term we might be able to bring back the colors)

**Pull request merged**
